File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length < 2)
        {
            return nums[0];
        }
        int m = robb(nums);
        int n = robb1(nums);
        return Math.max(m, n);
    }
    
    public int robb(int[] nums)
    {
        int[] arr = new int[nums.length];
        for (int i = 0;i<nums.length-1;i++)
        {
            arr[i] = nums[i];
        }
        arr[nums.length-1] = 0;
        for (int i = arr.length-3;i>-1;i--)
        {
            arr[i] = Math.max(arr[i]+arr[i+2], arr[i+1]);
        }
        return arr[0];
    }
    
    public int robb1(int[] nums)
    {
        int[] arr = new int[nums.length];
        for (int i = 1;i<nums.length;i++)
        {
            arr[i-1] = nums[i];
        }
        arr[nums.length-1] = 0;
        for (int i = arr.length-3;i>-1;i--)
        {
            arr[i] = Math.max(arr[i]+arr[i+2], arr[i+1]);
        }
        return arr[0];
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) < 4:
            return max(nums)
        def rob1():
            arr = [0] * (len(nums) - 1)
            arr[0] = nums[0]
            arr[1] = max(arr[0], nums[1])
            for i in range(2, len(nums) - 1):
                arr[i] = max(arr[i-1], arr[i-2] + nums[i])
            return arr[-1]
        def rob2():
            arr = [0] * (len(nums) - 1)
            arr[0] = nums[1]
            arr[1] = max(arr[0], nums[2])
            for i in range(3, len(nums)):
                arr[i-1] = max(arr[i-2], arr[i-3] + nums[i])
            return arr[-1]
        return max(rob1(), rob2())